### PR TITLE
Spider Revolution, Part 1: Laying the Groundwork

### PR DESCRIPTION
**Pull request by DeityLink**
image

imageimage
image

imageimageimage

:cl:
* experiment: Player-controlled Giant Spiders now have some sort of color-matrix powered night vision that lets them see quite well in the dark while still being able to identify lit areas, and therefore to seek lights to destroy. Might refine later. Feedback welcome.
* rscadd: Player-controlled Giant Spiders have a brand new HUD, featuring a health indicator, a fire warning, and a zone selection (for choosing which limb to target on humans). Spiderlings also have an indicator for their growth, while Nurse have indicators for both how many eggs they can lay, and how many spiders there are in total, as well as how many are needed before they can evolve into a Queen.
* rscadd: Player-controlled spiderlings are now able to ventcrawl (turns out they couldn't until now, despite NPC spiderlings being one of the only mob actually having an AI that lets it ventcrawl)
* rscadd: Player-controlled Giant Spiders now have spells for all their abilities that can be cast by players should they end up controlling one. This includes Spiderlings growing up (they get to choose whether to become Guard, Nurse or Hunter), as well as Nurse spinning webs, cocoons, laying eggs, and evolving into queens
* tweak: Spider hunters are now twice faster, still slower than carps though.
* rscadd: Changed the requirements and logic behind spider nurses evolving into queens. Before nurses would instantly become queens if there were less than 5 queens in the world. Now there has to be at least 10 spiders total (including spiderlings) per queen in the world for an NPC nurse to evolve. It will then have a 10% chance per second to evolve. Player-controlled nurses only require 6 spiders per queen, ensuring that NPC nurses shouldn't cuck players out of becoming queens. As a side-effect, non-queen nurses will become more common once again.
* rscadd: Spider queen projectiles no longer deal 5 brute damage, instead they root mobs to the floor just like blood nails do. As for those, resisting or clicking the webs will let you break out after one or two seconds. The queen can fire her projectile about every three or four seconds at best (BYOND time is uncertain).
* bugfix: Fixed the giant spider queen's projectile being invisible
* bugfix: Fixed NPC spider nurses laying two egg clusters when they should only be able to lay one, causing them to birth twice as many spiderlings as normal.

### CODE REVIEW

**Comment by Eneocho** [comment]
>
>It will then have a 10% chance per second to evolve.

Per actual second or is it byond tick?

**Comment by DeityLink** [comment]
byond tick are 1/10th of a second. If you want the technical detail the chance is rolled every Life() proc. Even though it's a bit more than a second I figured players would understand it better.

**Comment by Eneocho** [comment]
Spiders in general do not see well in the dark, but instead locate their prey by sound and vibration. I would change the visual overlay to a more sonar-like version. In the dark, you can see your prey, but the same way a Derilect MoMMI sees beings, as staticky silhouettes. Meaning there would be a choice between keeping lights on, so you can appropiately determine threats and prey, or break lights, so the crew can't see well, but neither do you.

**Comment by DeityLink** [comment]
Those are giant spess spiders, not your common house spiders. They've grown accustomed to the darkness of space, their eyes are so shiny because of how much light their reflect, just like cats'. Anyway here's your lore explanation.

From an actual gameplay standpoint I'd rather see who I'm attacking, the MoMMI filter sucks in my opinion but it makes sense for them because they're not supposed to discriminate between carbons due to their laws. No such things for spiders. And there's another flaw: unlike player-controlled xenos, player-controlled spiders are meant to be played alongside AI-controlled spiders, who ARE gonna break the lights anyway. By making lit areas slightly "too" bright I give players an incentive to mimic that behaviour and break lights.

**Comment by ghost** [comment]
\>two seconds to break out
\>four seconds between shots
terrifying

Also, the new calculation for queens sounds like it can be used to spider farm hardcore if it checks for spiderlings too.

**Comment by GreatBigFailure** [comment]
Neat
A lot of work for a, currently, very specific scenario but this does open doors for more ghost roles or something. Great work.

**Comment by DeityLink** [comment]
@sumireiro-sora keep in mind that spider farms are technically nerfed by 50% since now nurses can lay only one egg cluster per cocoon'd mob instead of two. But yes that is obviously the plan for player-controlled spiders:

Player-Nurse tries to cocoon as many mobs as possible, including monkeys and pets, while player-guards and player-hunters try to protect them, and NPC-spiders do their thing breaking lights and annoying the crew. I'm not completely set on how it'd play out but the idea would probably be something like:

1. up to 3 players get to spawn on top of a random vent as spiderlings during player infestations.
2. whenever one of them die, they get to take control of another spiderling until there are no spiderlings left on Z:1

I think there really shouldn't be more than 1-3 player controlled spiders among the swarm. But that'll be for another PR.

**Comment by 49bluesun** [comment]
But when do we get security spider riders to enforce the law?

Even if it doesn't end up standing out much from benos, a good looking first step to a new antagaroo

**Comment by DeityLink** [comment]
I'd say it stand out by being a lot less threatening. A single crew member armed with a toolbox can solo any spider except maybe a guard (or obviously a queen) while a single xenos can knock you down in one click and pull you into the rape cave.

The "player jumping to another spiderling" mechanic will allow us to still have relatively "weak" spiders individually, without it being too frustrating for the spider player.

**Comment by ghost** [comment]
@DeityLink I don't mean for the player controlled spiders, I mean in general. I'm assuming from the changelog that these changes are for all spiders, not just for a gamemode, in which case you'll definitely see people farming NPC queens by feeding them monkeymen/cargo order mobs. The buildup is slower, but the payoff is a lot better now, especially cause the sticky webs feel like a pretty substantial buff.

**Comment by DeityLink** [comment]
@sumireiro-sora but farming queens is easier in the current state of things 🤔 

As things stand, any nurse that spawns will become a Queen if there are less than 5 alive (that's why Roid engineering always have 3 queens at roundstart despite only 1 queen being mapped)

If you want more than 5 NPC queens after the update, you'll need to breed at least 60 spiders which is even slower to attain than before due to the egg laying fix (now you'll get 4-6 spiders per eaten mob instead of up to 8-12). And if you do farm 60+ spiders:

* not only do I not see how you could easily extract "only" the queen from wherever you breed them
* with this many spiders around, the swarm is gonna be deadly anyway and the queen buff won't be needed to just cause massive damage.

**Comment by drawsstuff** [comment]
MY TIME IS NOW

**Comment by TGFH** [comment]
Eneocho has the right idea with their suggestion about nightvision. The current nightvision seems to be near perfect and has no drawbacks whatsoever.

**Comment by drawsstuff** [comment]
> Eneocho has the right idea with their suggestion about nightvision. The current nightvision seems to be near perfect and has no drawbacks whatsoever.

So what, are you gonna nerf xenomorph's seeing in the dark as well?

**Comment by juanototo2** [comment]
Let spiders have their chance. Any new antagonist needs to have these first few rounds of being hilariously overpowered so they can be nerfed down and balanced later based on feedback.

**Comment by DeityLink** [comment]
> The current nightvision seems to be near perfect and has no drawbacks whatsoever

*seems* is the keyword here. It has a few display bugs that I probably won't be able to fix (dark objects becoming slightly transparent, actually transparent objects such as windows becoming fully opaque). Honestly I'll refrain from changes until players have had the chance of trying it out and giving feedback. I'm sure even without an actual spider role ghosts can ask admins to be put inside a spiderling.

**Comment by GreatBigFailure** [comment]
Out of the big round-threatening plays an antag can do I find mob loosing to be one of the more fun for the victims. It's never an immediate death. You can at least run away or hide, call for help, etc. It also often doesn't result in the station being so destroyed you might as well shuttle call anyway if the crew does come out on top.

Now all it needs is a non-antag reason to farm them.

**Comment by juanototo2** [comment]
> 
> 
> Out of the big round-threatening plays an antag can do I find mob loosing to be one of the more fun for the victims. It's never an immediate death. You can at least run away or hide, call for help, etc. It also often doesn't result in the station being so destroyed you might as well shuttle call anyway if the crew does come out on top.
> 
> Now all it needs is a non-antag reason to farm them.

The way TEEGEE did it, for xenomorphs at least, was allowing you to kill them, harvest their organs, and put their organs inside people so now Shaun Sciencemain can spit neurotoxin, make alien weeds, speak in alien chat, and use a few other alien abilities I don't quite remember. What would the spider organs for each spider type do, and how would you do it? Random chance to get while butchering?

**Comment by drawsstuff** [comment]
> > Out of the big round-threatening plays an antag can do I find mob loosing to be one of the more fun for the victims. It's never an immediate death. You can at least run away or hide, call for help, etc. It also often doesn't result in the station being so destroyed you might as well shuttle call anyway if the crew does come out on top.
> > Now all it needs is a non-antag reason to farm them.
> 
> The way TEEGEE did it, for xenomorphs at least, was allowing you to kill them, harvest their organs, and put their organs inside people so now Shaun Sciencemain can spit neurotoxin, make alien weeds, speak in alien chat, and use a few other alien abilities I don't quite remember. What would the spider organs for each spider type do, and how would you do it? Random chance to get while butchering?

Nothing because we arent fucking TG nor is this the pr for that

**Comment by Eneocho** [comment]
> 
>What would the spider organs for each spider type do, and how would you do it? Random chance to get while butchering? 

Make it so you can only remove one special organ, you get that organ but can't remove the others on that spider. If you remove the fangs and gland, you cant remove anything else, if you remove the eyes, you get 4 eyes to transplant (8 eyes per spider, only two eye sockets per human).
I'd make it so most transplants give a power/ability/skill but have SOME degree of incompatibility.

You could remove the poison gland and fangs from the spider and implant them via teeth surgery so now Med McMedmain can spit neurotoxin or whatever, but as a side effect, the poison glands paralize the vocal cords, so transplanteds are stronger, but can't speak/speak poorly (Lisp and No vowels? A phrase like: "Can I get a bag of holding please?" becomes "C_n _ g_t _ b_g _f h_ld_ng pl__th_?")
Eye transplants for that glorious nigh-perfect nightvision, but stronger effects from flashes/flashbangs (Flashblinded longer, twice as much?, double stun length). 
Queen Stinger would be an ass transplant. You can now shoot web but can't sit on anything, as if you had been arse n'ath'd.

Now, keep in mind I haven't played on TG so I have no idea if this is what you meant.

File: code/modules/mob/living/simple_animal/hostile/giant_spider/effects.dm [hunk 1]
```diff
@@ -180,4 +181,45 @@
...
+	if(isliving(A) && !isspace(T))//can't nail people down unless there's a turf to nail them to.
+		stuck_to = A
+		visible_message("<span class='warning'>\the sticky ball splatters over \the [A]'s legs, sticking them to \the [T].</span>")
+		lock_atom(A, /datum/locking_category/buckle)
+
+	spawn(duration)
+		qdel(src)
+
+/obj/effect/overlay/stickyweb/attack_hand(var/mob/user)
+	if (do_after(user,src,15))
```
Comment (Kurfursten) [inline code/modules/mob/living/simple_animal/hostile/giant_spider/effects.dm]: 1.5 SECONDS

also it seems weird that you can free yourself by hand but not destroy with a weapon?

File: code/modules/mob/living/simple_animal/hostile/giant_spider/effects.dm [hunk 2]
```diff
@@ -180,4 +181,45 @@
...
+	name = "sticky web"
+	desc = "A mess of sticky strings."
+	icon = 'icons/effects/effects.dmi'
+	icon_state = "stickyweb"
+	anchored = 1
+	density = 0
+	plane = ABOVE_HUMAN_PLANE
+	layer = CLOSED_CURTAIN_LAYER
+	var/atom/stuck_to = null
+	var/duration = 100
```
Comment (Kurfursten) [inline code/modules/mob/living/simple_animal/hostile/giant_spider/effects.dm]: 10 SECONDS

**Review by Kurfursten — COMMENTED** [review]
Had a look over every file except `giant_spider/nurse.dm` and `giant_spider/spiderling.dm`

if anyone else wants to do review, you only need to look over those 2 files

**Comment by drawsstuff** [comment]
MERGE WHENST

File: code/modules/mob/living/simple_animal/hostile/giant_spider/nurse.dm [hunk 3]
```diff
@@ -63,53 +126,65 @@
...
+		C.pixel_y = cocoon_target.pixel_y
+		for(var/mob/living/M in C.loc)
+			if(istype(M, /mob/living/simple_animal/hostile/giant_spider))
+				continue
+			for (var/atom/movable/AM in M.locked_atoms)
+				M.unlock_atom(AM)
+			if (M.locked_to)
+				M.locked_to.unlock_atom(M)
+			large_cocoon = 1
+			if(M.getCloneLoss() < 125)
```
Comment (ShiftyRail) [inline code/modules/mob/living/simple_animal/hostile/giant_spider/nurse.dm]: Make `1.25*M.maxHealth`

File: code/modules/mob/living/simple_animal/hostile/giant_spider/nurse.dm [hunk 4]
```diff
@@ -134,12 +210,35 @@
...
+		L.take_overall_damage(damage,0)
+		to_chat(L, "<span class='danger'>Resist or click the webs on your legs to free yourself.</span>")
+
 	if(!(locate(/obj/effect/spider/stickyweb) in get_turf(src)))
 		new /obj/effect/spider/stickyweb(get_turf(src))
+
+	bullet_die()
 	qdel(src)
+
+/obj/item/projectile/web/bump_original_check()
```
Comment (ShiftyRail) [inline code/modules/mob/living/simple_animal/hostile/giant_spider/nurse.dm]: This is one of the most confusing procs I've seen in a while, please explain what it does in a comment

File: code/modules/spells/spider_spells.dm [hunk 5]
```diff
@@ -0,0 +1,112 @@
+
+///////////////////////////////////SPIDERLING///////////////////////////////////////
+
+/spell/spiderling_evolution
```
Comment (ShiftyRail) [inline code/modules/spells/spider_spells.dm:4]: Need to add an `user_type` variable to these.

**Review by ShiftyRail — COMMENTED** [review]
I'm probably missing on some stuff but it looks ok

File: code/modules/spells/spider_spells.dm (hunk 5 above)
Comment (DeityLink) [reply code/modules/spells/spider_spells.dm:4]: didn't I do that 3 lines below?

**Review by ShiftyRail — APPROVED** [review]
In the best of all possible worlds I would ask you to make this matrix editing thing slightly less "dirty", but what you did is good enough

**Pull request merged**